runtime/lua/vim/secure.lua

```diff
@@ -86,21 +109,85 @@ function M.read(path)
     trust[fullpath] = hash
   end
 
-  do
-    local f, err = io.open(vim.fn.stdpath('state') .. '/trust', 'w')
-    if not f then
-      error(err)
+  write_trust(trust)
+
+  return contents
+end
+
+--- Update the trust status of file at {path} in the trust database at
+--- $XDG_STATE_HOME/nvim/trust.
+---
+---@param opts (table):
+---    - action (string): "allow" to add a file to the trust database and trust it,
+---      "deny" to add a file to the trust database and deny it,
+---      "remove" to remove file from the trust database
+---    - path (string|nil): Path to a file to update. If nil, must provide bufnr
+---    - bufnr (number|nil): Buffer number to update. If nil, must provide path.
+---      For "allow", must provide bufnr.
+---@return (boolean, string) success, msg:
+---    - true and full path of target file if operation was successful
+---    - false and error message on failure
+function M.trust(opts)
+  vim.validate({
+    path = { opts.path, 's', true },
+    bufnr = { opts.bufnr, 'n', true },
+    action = {
+      opts.action,
+      function(m)
+        return m == 'allow' or m == 'deny' or m == 'remove'
+      end,
+      [["allow" or "deny" or "remove"]],
+    },
+  })
+
+  local path = opts.path
+  local bufnr = opts.bufnr
+  local action = opts.action
+
+  if path and bufnr then
+    return false, 'path and bufnr are mutually exclusive'
+  end
+
+  local fullpath
+  if path then
+    fullpath = vim.loop.fs_realpath(vim.fs.normalize(path))
+  else
+    local bufname = vim.api.nvim_buf_get_name(bufnr)
+    if bufname == '' then
+      return false, 'buffer is not associated with a file'
+    end
+    fullpath = vim.loop.fs_realpath(vim.fs.normalize(bufname))
+  end
+
+  if not fullpath then
+    return false, string.format('invalid path: %s', path)
+  end
+
+  local trust = read_trust()
+
+  if action == 'allow' then
+    if not bufnr then
+      return false, 'allow operates on buffers - bufnr is required'
     end
 
-    local t = {}
-    for p, h in pairs(trust) do
-      t[#t + 1] = string.format('%s %s\n', h, p)
+    local contents = table.concat(vim.api.nvim_buf_get_lines(bufnr, 0, -1, false))
+    local hash = vim.fn.sha256(contents)
+
+    if trust[fullpath] ~= hash then
+      trust[fullpath] = hash
+    end
+  elseif action == 'deny' then
+    if trust[fullpath] ~= '!' then
+      trust[fullpath] = '!'
+    end
+  elseif action == 'remove' then
+    if trust[fullpath] then
+      trust[fullpath] = nil
     end
```

Comment:
```suggestion
    trust[fullpath] = hash
  elseif action == 'deny' then
    trust[fullpath] = '!'
  elseif action == 'remove' then
    trust[fullpath] = nil
```

Reply:
Resolved in 27f98a78c168ff895c2134c27350ce57e7cdbcfe